### PR TITLE
Remove pandas.core.index.datetimelike from MyPy Blacklist 

### LINKED ISSUE #25882: Typing Cleanup - Remove Blacklisted Items
#25844 introduced a blacklist of modules in the mypy.ini config file which currently throw errors when analyzed with mypy. We would ideally be able to whittle these down (save exclusion of the test folder(s), which I think is fine).

I'm not sure there is one easy way to split it up as the degree of difficulty may vary wildly. PRs from the community to clean up one to a few of the blacklisted modules at a time would certainly be welcome!

Here's the list of modules for reference:

- [X] pandas._config.config (#25899)
- [X] pandas._version (#25949)
- [X] pandas.compat (#26251)
- [X] pandas.compat.numpy.function
- [X] pandas.compat.pickle_compat (#25991)
- [X] pandas.core.accessor (#26146)
- [x] pandas.core.api
- [X] pandas.core.apply
- [X] pandas.core.arrays.array_  (#26065)
- [X] pandas.core.arrays.datetimelike  (#26065)
- [X] pandas.core.arrays.integer  (#26065)
- [X] pandas.core.arrays.interval (#26065)
- [X] pandas.core.arrays.numpy_ (#26065)
- [X] pandas.core.arrays.period (#26065)
- [X] pandas.core.arrays.sparse (#26065)
- [X] pandas.core.arrays.timedeltas (#26065)
- [x] pandas.core.base
- [x] pandas.core.computation.expr
- [x] pandas.core.computation.ops
- [x] pandas.core.computation.pytables
- [x] pandas.core.config (#26048)
- [x] pandas.core.config_init
- [X] pandas.core.dtypes.dtypes (#26028)
- [X] pandas.core.dtypes.missing (#26028)
- [X] pandas.core.frame
- [X] pandas.core.generic
- [X] pandas.core.groupby.generic (#26089)
- [X] pandas.core.groupby.groupby (#26089)
- [X] pandas.core.groupby.ops (#26089)
- [x] pandas.core.indexes.base (#26315)
- [x] pandas.core.indexes.datetimelike
- [ ] pandas.core.indexes.datetimes (#26315)
- [x] pandas.core.indexes.period (#26315)
- [x] pandas.core.indexes.timedeltas (#26315)
- [x] pandas.core.indexing
- [x] pandas.core.internals.blocks
- [x] pandas.core.ops
- [X] pandas.core.panel (DEPRECATED - do not update)
- [x] pandas.core.resample
- [X] pandas.core.reshape.concat (#26019)
- [x] pandas.core.reshape.merge
- [x] pandas.core.reshape.reshape
- [X] pandas.core.reshape.tile (#26019)
- [X] pandas.core.series
- [X] pandas.core.sparse.frame
- [X] pandas.core.strings (#25902)
- [x] pandas.core.util.hashing
- [x] pandas.core.window
- [x] pandas.io.clipboards (#26048)
- [x] pandas.io.feather_format (#25939)
- [x] pandas.io.formats.css (#25900)
- [x] pandas.io.html (#25939)
- [x] pandas.io.json.json (#25939)
- [x] pandas.io.json.normalize (#25939)
- [x] pandas.io.json.table_schema (#25939)
- [x] pandas.io.packers (#25939)
- [x] pandas.io.parquet (#25939)
- [x] pandas.io.pytables
- [x] pandas.io.stata (#25940)
- [X] pandas.plotting._core (#26021)
- [X] pandas.tseries.frequencies (#26014)
- [X] pandas.tseries.holiday (#26014)
- [X] pandas.tseries.offsets (#26014)
- [x] pandas.util._doctools
- [x] pandas.util.testing

### PR DESCRIPTION
**Pull request by makbigc**
- [x] closes #25882 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -69,7 +70,6 @@ class DatetimeIndexOpsMixin(ExtensionOpsMixin):
     _resolution = cache_readonly(DatetimeLikeArrayMixin._resolution.fget)
```
Comment (WillAyd) [inline pandas/core/indexes/datetimelike.py]: So it looks like these `cache_readonly` calls are what's tripping up mypy. @jbrockmendel might be able to offer some insight here

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the PR! Looks like there's an isort failure in CI

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: Is the problem with isort or mypy?

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: If its mypy then would it be that it wants something like `# type: cache_readonly`?

Comment (WillAyd) [reply pandas/core/indexes/datetimelike.py]: It's a mypy failure. Not actually with the cache_readonly as much as accessing the `.fget` attribute of the properties.

What's the intention of this code? Is passing fget from the property to cache_readonly faster than just accessing the property?

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: > What's the intention of this code? Is passing fget from the property to cache_readonly faster than just accessing the property?

`cache_readonly behaves like `property`, expecting to get a unary method as its argument.  Does it even work if you don't make the `.fget` explicit?

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -134,7 +134,7 @@ def values(self) -> np.ndarray:
         # Note: PeriodArray overrides this to return an ndarray of objects.
         return self._data._data
 
-    @property
+    @property  # type: ignore
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
Can you merge master?

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (makbigc) [reply pandas/core/indexes/datetimelike.py]: This failure is probably due to mypy. [https://github.com/python/mypy/issues/6185](url)

Deleting `.fget` run into other errors.

**Comment by vaibhavhrt** [comment]
@makbigc some of the errors in this file has been solved by #26404 
You can go ahead and fix(or ignore) the remaining.

**Comment by WillAyd** [comment]
Can you merge master? Think this will be the last remaining module from the blacklist

**Comment by jreback** [comment]
can you merge master

**Comment by makbigc** [comment]
mypy has the following error for the annotated value function: 

`pandas/core/indexes/datetimelike.py:139: error: "None" has no attribute "_data"`

I add annotation ` # type: DatetimeArray ` to `_data`. But mypy still has the same error.
https://github.com/pandas-dev/pandas/blob/8154efb0c1a64295cf54e00025b4ab09bcd02752/pandas/core/indexes/datetimes.py#L327

Do you have any better way?

File: pandas/core/indexes/datetimelike.py [hunk 3]
```diff
@@ -130,11 +134,11 @@ def _ndarray_values(self):
     # Abstract data attributes
 
     @property
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: why this change?

Comment (makbigc) [reply pandas/core/indexes/datetimelike.py]: mypy has the following error for the annotated value function: 

`pandas/core/indexes/datetimelike.py:139: error: "None" has no attribute "_data"`

I add annotation ` # type: DatetimeArray ` to `_data`. But mypy still has the same error.
https://github.com/pandas-dev/pandas/blob/8154efb0c1a64295cf54e00025b4ab09bcd02752/pandas/core/indexes/datetimes.py#L327

I guess that mypy mixed up the instance attribute `_data` with the class attributes `_data` which is set `None` in the beginning of DatetimeIndexOpsMixin.

Comment (jreback) [reply pandas/core/indexes/datetimelike.py]: cc @WillAyd

Comment (WillAyd) [reply pandas/core/indexes/datetimelike.py]: Sorry missed this ping. This could be a similar conversation to https://github.com/pandas-dev/pandas/pull/26518/files#r287573810

Comment (WillAyd) [reply pandas/core/indexes/datetimelike.py]: I think a better solution is to provide the appropriate type to `_data`. Can you create a `DatetimeLikeArray` in pandas._typing which should be something like `DatetimeLikeArray = TypeVar('DatetimeLikeArray', DatetimeArray, PeriodArray, TimedeltaArray)` and assign that type to `_data` in the class? I think that should resolve.

cc @jbrockmendel in case he has other insights on the types here

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
@makbigc can you merge master and address last comment? This should close out the reference issue (updated OP to reflect)

**Comment by makbigc** [comment]
I can't figure out why DatetimeArray, PeriodArray and TimedeltaArray cannot be imported inside _typing.py

`from pandas.core.arrays import DatetimeArray, PeriodArray, TimedeltaArray`

File: pandas/_typing.py [hunk 4]
```diff
@@ -11,12 +11,17 @@
 from pandas.core.dtypes.generic import (
     ABCExtensionArray, ABCIndexClass, ABCSeries, ABCSparseSeries)
 
+from pandas.core.arrays import DatetimeArray, PeriodArray, TimedeltaArray
+
```
Comment (jreback) [inline pandas/_typing.py]: don’t try to import these use a quoted version

Comment (makbigc) [reply pandas/_typing.py]: The following is the import error in the failed tests:
`ImportError: cannot import name '_INT64_DTYPE'`

The name of the module _typing.py is started with underscore. In PEP 8 ([https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles](url) ), it states

> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.

Should we remove underscore from the name _typing? Or something else?

Comment (jreback) [reply pandas/_typing.py]: you can’t import these names as they would cause circular imports 
and don’t need to just use ‘DatetimeArray’ and such

Comment (makbigc) [reply pandas/_typing.py]: `_INT64_DTYPE` is just an alias, i.e.,  `_INT64_DTYPE = np.dtype(np.int64)`. The import of numpy and the definition of `_INT64_DTYPE` are in the same file `pandas/core/dtype/common.py`.

I did a little experiment. Open a file called 'typing.py' (without underscore prefix) in the same directory as '_typing.py'. One can import `DatetimeArray` in `typing.py`.

Importing `PeriodArray` and `TImedeltaArray` also run into the same `ImportError`. Without them, one cannot define the mypy object `TypeVar`.
`DatetimeLikeArray = TypeVar('DatetimeLikeArray', DatetimeArray, PeriodArray, TimedeltaArray)`.

Comment (makbigc) [reply pandas/_typing.py]: @WillAyd Would you tell me your opinion? Should we change the filename of _typing.py?

**Comment by WillAyd** [comment]
No I don’t think this needs to be renamed. I don’t have a problem using the TYPE_CHECKING constant in that module alone if it resolves circular imports.

**Comment by makbigc** [comment]
Yes, the problem is due to the circular import.
`common` -> `_typing` -> `datetimelikey` -> `common`
I can't resolve this circular import and datetimelike.py is the last exception in mypy.ini. So I close this PR for others picking it up.

**Comment by WillAyd** [comment]
Doesn’t using the TYPE_CHECKING constant from the stdlib typing module in pandas._typing resolve the circular import?

I think this is really close!

**Comment by makbigc** [comment]
@WillAyd Thanks for your reply. I have added the following in `_typing.py`
```
if TYPE_CHECKING:
    from pandas.core.arrays.datetimes import DatetimeArray
    from pandas.core.arrays.period import PeriodArray
    from pandas.core.arrays.timedeltas import TimedeltaArray
    DatetimeLikeArray = TypeVar('DatetimeLikeArray', DatetimeArray,
                                PeriodArray, TimedeltaArray)
```
and added the following to datetimelike.py
```
if TYPE_CHECKING:
    from pandas._typing import DatetimeLikeArray
```
The circular import is solved. But mypy cannot find the DatetimeLikeArray!
`pandas/core/indexes/datetimelike.py:67: error: Invalid type "pandas.DatetimeLikeArray"`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @WillAyd merge when ready.

**Comment by WillAyd** [comment]
Thanks @makbigc !
